avogadro/io/dcdformat.cpp

```diff
@@ -0,0 +1,432 @@
+  limitations under the License.
+
+******************************************************************************/
+
+#include "dcdformat.h"
+#include "struct.h"
+
+#include <avogadro/core/elements.h>
+#include <avogadro/core/molecule.h>
+#include <avogadro/core/unitcell.h>
+#include <avogadro/core/utilities.h>
+#include <avogadro/core/vector.h>
+
+#include <cmath>
+#include <iomanip>
+#include <istream>
+#include <ostream>
+#include <sstream>
+#include <string>
+
+using std::endl;
+using std::getline;
+using std::map;
+using std::pair;
+using std::string;
+using std::to_string;
+using std::vector;
+
+namespace Avogadro {
+namespace Io {
+
+using Core::Array;
+using Core::Atom;
+using Core::Elements;
+using Core::lexicalCast;
+using Core::Molecule;
+using Core::split;
+using Core::trimmed;
+using Core::UnitCell;
+
+#ifndef _WIN32
+using std::isalpha;
+#endif
+
+#define DCD_EOF -1
+#define DCD_MAGIC 84
+#define DCD_IS_CHARMM 0x01
+#define DCD_HAS_4DIMS 0x02
+#define DCD_HAS_EXTRA_BLOCK 0x04
+
+int swap_integer(int inp)
+{
+  return (((inp << 24) & 0xff000000) | ((inp << 8) & 0x00ff0000) |
+          ((inp >> 8) & 0x0000ff00) | ((inp >> 24) & 0x000000ff));
+}
+
+char swap_endian(char endian)
+{
+  if (endian == '>')
+    return '<';
+  else
+    return '>';
+}
+
+DcdFormat::DcdFormat() {}
+
+DcdFormat::~DcdFormat() {}
+
+bool DcdFormat::read(std::istream& inStream, Core::Molecule& mol)
+{
+  char endian = '>', buff[BUFSIZ], fmt[BUFSIZ], raw[84];
+  char* remarks;
+  double DELTA;
+  int magic, charmm, NSET, ISTART, NSAVC, NAMNF, NTITLE, lenRemarks, NATOMS,
+    blockSize;
+
+  // Determining size of file
+  inStream.seekg(0, inStream.end);
+  int fileLen = inStream.tellg();
+  inStream.seekg(0, inStream.beg);
+
+  // Reading magic number
+  snprintf(fmt, sizeof(fmt), "%c1i", endian);
+  inStream.read(buff, struct_calcsize(fmt));
+  struct_unpack(buff, fmt, &magic);
+  if (magic != DCD_MAGIC) {
+    magic = swap_integer(magic);
+    endian = swap_endian(endian);
+    if (magic != DCD_MAGIC) {
+      appendError("File does not start with magic number 84.");
+      return false;
+    }
+  }
+
+  // CORD
+  snprintf(fmt, sizeof(fmt), "%c%ds", endian, magic);
+  inStream.read(buff, struct_calcsize(fmt));
+  struct_unpack(buff, fmt, raw);
+  if (raw[0] != 'C' || raw[1] != 'O' || raw[2] != 'R' || raw[3] != 'D') {
+    appendError("Keyword CORD not found.");
+    return false;
+  }
+
+  // Determining whether the trajectory file is from CHARMM or not
+  if (*((int*)(raw + 80)) != 0) {
+    charmm = DCD_IS_CHARMM;
+    if (*((int*)(raw + 44)) != 0)
+      charmm |= DCD_HAS_EXTRA_BLOCK;
+
+    if (*((int*)(raw + 48)) == 1)
+      charmm |= DCD_HAS_4DIMS;
+  } else {
+    charmm = 0;
+  }
+
+  // number of fixed atoms
+  NAMNF = *((int*)(raw + 36));
+
+  // DELTA (timestep) is stored as a double with X-PLOR but as a float with
+  // CHARMM
+  if ((charmm)&DCD_IS_CHARMM) {
+    float ftmp;
+    ftmp = *((float*)(raw + 40));
+
+    DELTA = (double)ftmp;
```

Comment:
Avoid the use of C style casts, prefer static_cast<> where reasonable, other C++ casts where needed.